python/lsst/cp/pipe/ptc/cpPtcSolve.py

```diff
@@ -1077,14 +1110,168 @@ def _makeZeroSafe(self, array, substituteValue=1e-9):
+            `PhotonTransferCurveDatase`.
+        """
+        # For FULLCOVARIANCE model fit
+        matrixSideFit = dataset.covMatrixSideFullCovFit
+        nanMatrixFit = np.empty((matrixSideFit, matrixSideFit))
+        nanMatrixFit[:] = np.nan
+
+        for amp in dataset.ampNames:
+            lenInputTimes = len(dataset.rawExpTimes[amp])
+            listNanMatrixFit = np.empty((lenInputTimes, matrixSideFit, matrixSideFit))
+            listNanMatrixFit[:] = np.nan
+
+            dataset.covariancesModel[amp] = listNanMatrixFit
+            dataset.aMatrix[amp] = nanMatrixFit
+            dataset.bMatrix[amp] = nanMatrixFit
+            dataset.noiseMatrix[amp] = nanMatrixFit
+
+        def errFunc(p, x, y):
+            return ptcFunc(p, x) - y
+
+        for i, ampName in enumerate(dataset.ampNames):
+            meanVecOriginal = dataset.rawMeans[ampName].copy()
+            varVecOriginal = dataset.rawVars[ampName].copy()
+            varVecOriginal = self._makeZeroSafe(varVecOriginal)
+
+            # These must be sorted for the given amplifier.
+            meanVecSort = np.argsort(meanVecOriginal)
+            meanVecSorted = meanVecOriginal[meanVecSort]
+            varVecSorted = varVecOriginal[meanVecSort]
+
+            # Make sure we have this properly sorted.
+            goodPoints = dataset.expIdMask[ampName].copy()
+            goodPoints = goodPoints[meanVecSort]
+
+            # Check if all points are bad from the previous fit.
+            initialExpIdMask = dataset.expIdMask[ampName].copy()
+
+            if not (goodPoints.any() and initialExpIdMask.any()):
+                msg = (f"SERIOUS: All points in goodPoints: {goodPoints} or "
+                       f"in initialExpIdMask: {initialExpIdMask} are bad."
+                       f"Setting {ampName} to BAD.")
+                self.log.warning(msg)
+                # Fill entries with NaNs
+                self.fillBadAmp(dataset, ampName)
+                continue
+
+            mask = goodPoints.copy()
+
+            # Compute the extended mask
+            preTurnoff = dataset.ptcTurnoff[ampName]
+            turnoffIdx = np.argwhere(meanVecSorted == preTurnoff)[0][0]
+            varianceAtTurnoff = varVecSorted[turnoffIdx]
+
+            # Add points up to some threshold below the variance of the turnoff
+            pointsToFit = (meanVecSorted >= preTurnoff) * \
+                          (meanVecSorted < preTurnoff+self.config.extendRollofSearchMaskSizeAdu) * \
+                          (varVecSorted <= varianceAtTurnoff)
+
+            # Retain the original mask below the turnoff
+            pointsToFit = np.logical_or(mask, pointsToFit)
+            if np.count_nonzero(pointsToFit) == np.count_nonzero(mask):
+                self.log.warning("Expanding fit to include saturation, but no points detected above "
+                                 f"initial computed PTC turnoff for amp {ampName}.")
```

Comment:
This warning is saying there's no points _above_ the original turnoff ... but we know that the original turnoff is already over-estimated.  So I'm not sure why this is a fatal problem?  (What happens for the highly vignetted amps?)